rfc/rfc-27/rfc-27.md

```diff
@@ -0,0 +1,409 @@
+                            "doc" : "Stores total null values for this column in the resepective data file"
+                         },
+                        {
+                            "name":"total_nans",
+                            "type":["long", "null"],
+                            "doc" : "Stores total Nan values for this column in the resepective data file"
+                         },
+                        {
+                            "name":"total_size_on_disk",
+                            "type":["long", "null"],
+                            "doc" : "Stores total size occupied by this column on disk corresponding to the resepective data file"
+                         },                                                                           
+                        {
+                            "name": "isDeleted",
+                            "type": "boolean",
+                            "doc": "True if this file has been deleted"
+                        }
+                    ]
+                }
+            ]
+        }
+```
+
+Column stats records hold all stats for the file. The key for the column stat record would be an
+encoded string as discussed earlier. 
+
+```
+key = base64_encode(hash64(column name) + hash64(partition name) + hash128(file path))
+key = base64_encode(hash64(column name) + "agg" + hash64(partition name))
+```
+
+While Hash based IDs have quite a few desirable properties in the context of Hudi index lookups, there is an impact
+on the column level schema changes though. Refer to [Schema Evolution](#Schema-Evolution) section for more details.
+
+#### Writer flow
+Let's walk through the writer flow to update column_stats partition in metadata table.
+
+1. Files partition - prepare records for adding // just calling out whats required in the context of column_stats 
+   partition. General files partition will be updated as usual to store file listing information.
+    - FileID => FileName mapping entries
+    - PartitionID => PartitionName entry, if not already exists
+    - Since these IDs are hash based IDs, no look up of prior usages is required here. If not, we need to know what was
+   the last assigned ID and then go about assigning new incremental/sequential IDs, which slows down the performance significantly
+2. Column_stats partition - prepare records for adding
+    - [ColumnID][PartitionID][FileID] => ColumnStat
+    - [ColumnId]"agg"[PartitionId] => ColumnStat
+    - This involves reading the base file footers to fetch min max and other stats to populate values for the record.
+d. Commit all these records to metadata table.
+
+We need to ensure we have all sufficient info in WriteStatus/Commit Metadata that gets passed to metadata writer for 
+every commit. Reading parquet footers and meta is unavoidable, but other than that, we should try to embed all other info 
+in the WriteStatus.
+
+### Index integrations with query engines
+
+#### How to apply query predicates in Hudi?
```

Comment:
This section will have to be revisited. With recent support for z-ordering, we can re-use the data skipping utils we already have in out code base. please wait to review this section in detail.

But high level flow will be as follows:
if z-ordering is enabled, at the end of z-order clustering, the replace commit will get applied to metadata table. So, we don't need to build any index as such for z-ordering as such. 
On the query side, we will rely on what column stats partition in metadata gives us. If z-order clustering was complete, column_stats partition will reflect it. If not, it will return the current state of things. So, query side integration may not need any branching on whether z-ordering is enabled or not and can just rely on the column_stats partition. 

One caveat:
But we can't get rid of z-order index completely though right away. If metadata table is not build out yet or has entered an inconsistent state and is not usable, we have to go the existing way of building an index at the end of z-order clustering. 

Hope this gives a high level idea of how we plan to do query side integration.

Reply:
have updated the rfc with this info.